pkg/server/apiextensions.go

```diff
@@ -33,135 +33,201 @@ import (
+		if err != nil {
+			// TODO return?
+			break
+		}
+		if clusterWorkspace.Spec.Type == "Universal" {
+			return p.universalCRDs
+		}
+	}
+
+	return sets.NewString()
+}
+
+// apiBindingAwareCRDLister is a CRD lister combines APIs coming from APIBindings with CRDs in a workspace.
+type apiBindingAwareCRDLister struct {
+	crdLister         apiextensionslisters.CustomResourceDefinitionLister
+	workspaceLister   tenancylisters.ClusterWorkspaceLister
+	apiBindingLister  apislisters.APIBindingLister
+	systemCRDProvider *systemCRDProvider
+}
+
+var _ apiextensionslisters.CustomResourceDefinitionLister = (*apiBindingAwareCRDLister)(nil)
 
 // List lists all CustomResourceDefinitions in the underlying store matching selector. This method does not
-// support scoping to logical clusters or workspace inheritance.
-func (c *inheritanceCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// support scoping to logical clusters or APIBindings.
+func (c *apiBindingAwareCRDLister) List(selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	return c.crdLister.ListWithContext(context.Background(), selector)
 }
 
-// ListWithContext lists all CustomResourceDefinitions in the logical cluster associated with ctx that match
-// selector. ClusterWorkspace API inheritance is also supported: if the ClusterWorkspace for ctx's logical cluster
-// has spec.inheritFrom set, it will aggregate all CustomResourceDefinitions from the ClusterWorkspace named
-// spec.inheritFrom with the CustomResourceDefinitions from the ClusterWorkspace for ctx's logical cluster.
-func (c *inheritanceCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+// ListWithContext lists all CustomResourceDefinitions that come in via APIBindings as well as all in the current
+// logical cluster.
+func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector labels.Selector) ([]*apiextensionsv1.CustomResourceDefinition, error) {
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	// Check for API inheritance
-	inheriting := false
-	inheritFrom := ""
-	orgName, ws, err := helper.ParseLogicalClusterName(cluster.Name)
+	var ret []*apiextensionsv1.CustomResourceDefinition
+	crdName := func(crd *apiextensionsv1.CustomResourceDefinition) string {
+		return crd.Spec.Names.Plural + "." + crd.Spec.Group
+	}
+	seen := sets.NewString()
+
+	org, workspace, err := helper.ParseLogicalClusterName(cluster.Name)
+	if err != nil {
+		return nil, fmt.Errorf("error determining workspace name from cluster name %q: %w", cluster.Name, err)
+	}
+
+	kcpSystemCRDs, err := c.systemCRDProvider.List(ctx, org, workspace)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving kcp system CRDs: %w", err)
+	}
+
+	for i := range kcpSystemCRDs {
+		ret = append(ret, kcpSystemCRDs[i])
```

Comment:
not equal to `ret = kcpSystemCRDs`?

Reply:
Yeah, but I needed to get all the names in `seen`. Can make your suggested change if you want

Reply:
either is fine. Just want to confirm I don't miss anything :)